### DISCUSSION TITLE
nested `FilterDSL` with generic type

### DISCUSSION THREAD

**Discussion by vnghia** [diesel-rs/diesel#3945] [Q&A] [answered]
```rust
use diesel::{
    dsl::{exists, Eq, Filter},
    expression::AsExpression,
    query_dsl::methods::FilterDsl,
    sql_types, ExpressionMethods,
};

mod schema {
    diesel::table! {
        users (id) {
            id -> Uuid,
        }
    }

    diesel::table! {
        posts (id) {
            id -> Uuid,
        }
    }
}

use schema::*;

// this compiles
pub fn get_user<U>(id: U) -> Filter<users::table, Eq<users::id, U>>
where
    U: AsExpression<sql_types::Uuid>,
    users::table: FilterDsl<Eq<users::id, U>>,
{
    users::table.filter(users::id.eq(id))
}

// this does not compile
pub fn get_posts<U>(id: U) -> Filter<posts::table, exists<Filter<users::table, Eq<users::id, U>>>>
where
    U: AsExpression<sql_types::Uuid>,
    users::table: FilterDsl<Eq<users::id, U>>,
    posts::table: FilterDsl<exists<Filter<users::table, Eq<users::id, U>>>>,
{
    posts::table.filter(exists(users::table.filter(users::id.eq(id))))
}

```

The constraint `users::table: FilterDsl<Eq<users::id, U>>` isn't taken into account and the compiler complaints that `FilterDsl<Eq<users::id, U>> is not implemented for users::table`.

I think it is more likely an issue with rust type system but I want to ask here to see if there is a diesel specific workaround beside using `BoxableExpression` since I want to do as many thing as possible in compile time.

I think the compiler expects something like:
```rust
pub fn get_posts<U>(id: U) -> Filter<posts::table, exists<Filter<users::table, Eq<users::id, U>>>>
where
    U: AsExpression<sql_types::Uuid>,
    users::table: FilterDsl<Eq<users::id, U>> + posts::table: FilterDsl<exists<Filter<users::table, Eq<users::id, U>>>>,
{
    posts::table.filter(exists(users::table.filter(users::id.eq(id))))
}
```

but that isn't valid rust syntax :/

**Comment by weiznich** [comment, marked as answer]
That looks like a bug in the rust compiler. It might be worth to report that upstream, although I believe there is not much that they can do about that. 
Unfortunately that means there is also not much we can do about that in diesel.

**Reply by vnghia** [reply]
thank you very much! I will try to have a minimal reproductible example without diesel and report to rust compiler team

**Reply by weiznich** [reply]
That would be great :+1: Feel free to ping me from the rust issue.
